### PR TITLE
include examples source code into assembly

### PR DESCRIPTION
**Pull request by rmetzger**
The source code for the examples is now placed inside
`/examples/scala-src` and `/examples/java-src`.

```
robert@robert-tower ...bin/stratosphere-0.5-SNAPSHOT/examples (git)-[exsrc] % find .
.
./stratosphere-java-examples-0.5-SNAPSHOT-EnumTrianglesWithDegrees.jar
./stratosphere-java-examples-0.5-SNAPSHOT-WebLogAnalysis.jar
./stratosphere-java-examples-0.5-SNAPSHOT-EnumTrianglesOnEdgesWithDegrees.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-ConnectedComponents.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-TPCHQuery3.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-WordCount.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-javadoc.jar
./scala-src
./scala-src/datamining
./scala-src/datamining/BatchGradientDescent.scala
./scala-src/datamining/LanczosSO.scala
./scala-src/datamining/KMeansForTest.scala
./scala-src/datamining/KMeans.scala
./scala-src/graph
./scala-src/graph/PageRank.scala
./scala-src/graph/EnumTrianglesOnEdgesWithDegrees.scala
./scala-src/graph/LineRank.scala
./scala-src/graph/ComputeEdgeDegrees.scala
./scala-src/graph/ConnectedComponents.scala
./scala-src/graph/TransitiveClosureRD.scala
./scala-src/graph/TransitiveClosureNaive.scala
./scala-src/graph/PageRankWithWeight.scala
./scala-src/wordcount
./scala-src/wordcount/WordCount.scala
./scala-src/wordcount/WordCountWithUserDefinedType.scala
./scala-src/wordcount/WordCountWithCount.scala
./scala-src/grabbag
./scala-src/grabbag/Grabbag.scala
./scala-src/relational
./scala-src/relational/WebLogAnalysis.scala
./scala-src/relational/RelationalQuery.scala
./java-src
./java-src/incremental
./java-src/incremental/pagerank
./java-src/incremental/pagerank/DeltaPageRankWithInitialDeltas.java
./java-src/triangles
./java-src/triangles/EnumTrianglesRdfFoaf.java
./java-src/triangles/EnumTrianglesWithDegrees.java
./java-src/triangles/io
./java-src/triangles/io/EdgeWithDegreesInputFormat.java
./java-src/triangles/io/TriangleOutputFormat.java
./java-src/triangles/io/EdgeWithDegreesOutputFormat.java
./java-src/triangles/io/EdgeInputFormat.java
./java-src/triangles/ComputeEdgeDegrees.java
./java-src/triangles/EnumTrianglesOnEdgesWithDegrees.java
./java-src/shortestpaths
./java-src/shortestpaths/PairwiseSP.java
./java-src/sort
./java-src/sort/terasort
./java-src/sort/terasort/TeraKey.java
./java-src/sort/terasort/TeraValue.java
./java-src/sort/terasort/TeraOutputFormat.java
./java-src/sort/terasort/TeraInputFormat.java
./java-src/sort/terasort/TeraDistribution.java
./java-src/sort/ReduceGroupSort.java
./java-src/sort/TeraSort.java
./java-src/kmeans
./java-src/kmeans/udfs
./java-src/kmeans/udfs/PointOutFormat.java
./java-src/kmeans/udfs/RecomputeClusterCenter.java
./java-src/kmeans/udfs/ComputeDistance.java
./java-src/kmeans/udfs/PointInFormat.java
./java-src/kmeans/udfs/CoordVector.java
./java-src/kmeans/udfs/FindNearestCenter.java
./java-src/kmeans/udfs/ComputeDistanceParameterized.java
./java-src/kmeans/KMeans.java
./java-src/kmeans/KMeansIterative.java
./java-src/kmeans/KMeansSampleDataGenerator.java
./java-src/util
./java-src/util/ConfigUtils.java
./java-src/connectedcomponents
./java-src/connectedcomponents/WorksetConnectedComponents.java
./java-src/wordcount
./java-src/wordcount/WordCountOptimized.java
./java-src/wordcount/AnonymousWordCount.java
./java-src/wordcount/WordCountAccumulators.java
./java-src/wordcount/WordCount.java
./java-src/relational
./java-src/relational/generator
./java-src/relational/generator/WebLogGenerator.java
./java-src/relational/WebLogAnalysis.java
./java-src/relational/TPCHQuery3.java
./java-src/pagerank
./java-src/pagerank/DiffL1NormConvergenceCriterion.java
./java-src/pagerank/PageRankStats.java
./java-src/pagerank/DotProductMatch.java
./java-src/pagerank/DanglingPageRankInputFormat.java
./java-src/pagerank/DanglingPageRank.java
./java-src/pagerank/DotProductCoGroup.java
./java-src/pagerank/PageWithRankOutFormat.java
./java-src/pagerank/PageRankStatsAggregator.java
./java-src/pagerank/ImprovedAdjacencyListInputFormat.java
./java-src/pagerank/SimplePageRank.java
./java-src/pagerank/LongArrayView.java
./java-src/pagerank/AsciiLongArrayView.java
./stratosphere-scala-examples-0.5-SNAPSHOT-KMeansForTest.jar
./stratosphere-java-examples-0.5-SNAPSHOT-TPCHQuery3.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-KMeans.jar
./stratosphere-java-examples-0.5-SNAPSHOT-DanglingPageRank.jar
./stratosphere-java-examples-0.5-SNAPSHOT-ConnectedComponents.jar
./stratosphere-java-examples-0.5-SNAPSHOT-KMeansIterative.jar
./stratosphere-java-examples-0.5-SNAPSHOT-TeraSort.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-ComputeEdgeDegrees.jar
./stratosphere-java-examples-0.5-SNAPSHOT-PairwiseSP.jar
./stratosphere-java-examples-0.5-SNAPSHOT-EnumTrianglesRdfFoaf.jar
./stratosphere-java-examples-0.5-SNAPSHOT-javadoc.jar
./stratosphere-scala-examples-0.5-SNAPSHOT-EnumTrianglesOnEdgesWithDegrees.jar
./stratosphere-java-examples-0.5-SNAPSHOT-WordCount.jar
./stratosphere-java-examples-0.5-SNAPSHOT-KMeans.jar
```

### CODE REVIEW

**Comment by StephanEwen** [comment]
Nice work!

Out of curiosity: You cannot directly load the directoty `./java-src`into an IDE, because the directory path `pagerank/DanglingPageRank.java` does not match the package (missing the prefix `eu.stratosphere.example...`). Would it make sense to retain the deep paths? On the pros, it would work easier to import it into an IDE, on the cons, it makes the file based browsing more tedious (deep paths).

What is your take on that?

**Comment by rmetzger** [comment]
If people want to inspect the code in an IDE, I would recommend to import the `stratosphere-examples` project as a maven project.
My idea for this was to give people a convenient way to access the source code, for example with a simple text editor.
The only feasible approach to allow this would be to change the package name. But I'm not aware of any maven-based solution for this.

**Comment by nullusDefectus** [comment]
As a potential user of Stratosphere, I'd rather it be easy to open the example source in an IDE. I can always ask the IDE to take me to the folder containing a source file (and that would be rare). 

In general, users' liking or disliking to a product is highly related to the effort they have to put on first contact. So, my view is that if users install the Ready to Run Package, they should be able to open the example source in Eclipse or other IDE, set the class path to Stratosphere "bin", and run the examples from the IDE without installing anything else, except for examples that illustrate using Stratosphere with Hadoop (which I think can be further addressed by separating the Hadoop examples from non-Hadoop examples).

**Comment by rmetzger** [comment]
Hi,
thank you for your feedback!

You are right. Exploring a complex software such as Stratosphre without IDE support is probably impossible!
I also agree with your second comment. We probably have 15-45 minutes to convince potential users of the greatness of our software.

I have however a different view on providing source code in the ready-to-run packages.

Lets say we have two different audiences:
a) Software Developers (Java, Scala)
b) System Administrators.

For a), the developers follow the [java quickstart guide](http://stratosphere.eu/quickstart/java.html) on the website. It will create a sample project skeleton with everything required to develop and debug a Stratosphere job on a computer. The only step to set this up is invoking one script and importing the files into eclipse! (they do not need the ready-to-run packages)
If a developer wants to access (or execute) the examples, they just have to open the class. (again, no ready-to-run binaries required! -- everything out of the IDE).
Once the developer has finished writing his Job and wants to execute it on the cluster, he has to download the "ready to run package" for the cluster. We provide the examples with the packages so that they can validate the correctness of the installation.

So to conclude the a)-case: Software developers will most likely have their first contact with Stratosphere using our quickstarts (the technology behind this are maven archetypes).

Users from category b) will download the ready to run packages, install Stratospehre on their local machine or cluster and run the examples. They can look at the webinterface or in the log files.

(We also have/had the Quickstart Virtual Machine for users that are not so familiar with Linux or are frighten by the installation)

It might be the case that I have the wrong assumption about our user's behavior.

**Comment by nullusDefectus** [comment]
I will be careful writing my observations (and ask for a wide berth), because everything I know about Stratosphere is limited to what I have read on its web site in the last week or so. You probably have been on the development project from its inception, so I will defer to you on the project's internals, goals, and priorities.

With that said, assuming the Ready to Run Package includes example source:
1. Is Maven is necessary to run the example source when users have only the Ready to Run Package? I can understand if Maven is recommended to view/build the complete Stratosphere source.
2. What are the impediments to users viewing/building/running the examples in an IDE _without_ using Maven when they have only the Ready to Run Package?

Here are some notes from a quick trial I did: I created an Eclipse project (I don't have Maven) and added reference to jars in Stratosphere's lib and lib_clients directories. I then added the example WordCount.java (extracted from Stratosphere source) to the project and the project compiled successfully, except the compiler did not approve of using the 4-arg constructor of class CsvOutputFormat. The program did compile when I changed the offending line to use the default constructor. (I did not have the time to investigate why the original constructor was unacceptable.)

So, at least superficially, it seems I am able to view/build the example source in Eclipse without using Maven.

Unfortunately, running WordCount caused an exception, but in the limited time I spent investigating it, I don't immediately see why this exception should occur (WordCount.TokenizeLine's ancestor AbstractFunction implements Seralized). I assume the exception is caused by incorrect order of import or references (or some such programmer silliness), but I am left with the following questions:
1. Are there Stratosphere dependencies I am not meeting in this trial? I really hope not.
2. Would using Maven eliminate both the compile-time and run-time errors? I doubt it, and hope the answer isn't "yes, it does, because it makes sure you have the latest everything.".
3. Is it necessary for the user to go through this process to get some instant gratification (which is key to user engagement)? I don't believe they should.

BTW, I spent close to 90 calendar minutes conducting this trial, which IMHO is rather high, especially if I don't yet have that instant gratification. :)

image

**Comment by rmetzger** [comment]
Hi,
I guess you've downloaded the 0.4 release (ready to run package) but used the current source code from the repository for your experiment ?
They are incompatible since the source code in the repo belongs to version 0.5-SNAPSHOT.

**Comment by rmetzger** [comment]
Now let me answer your questions
1. No, users do not need Maven to run the compiled examples inside the ready to run packages.
2. When you download Eclipse from eclipse.org, it already contains maven support. You do not have to install anything. (The same applies to IntelliJ). So most of the users already have or use maven, without knowing it.
   Maven is the de-facto standard for all other "big data" projects such as Hadoop, HBase, Giraph etc. So our users should be used to the typical maven commands (most importantly `mvn clean package -DskipTests`).

What you did should be possible if you do not mix the versions (its just an assumption, you might have found a bug). So yes, you can run Stratosphere without Maven only with the jars in the ready to run packages. 

```
except the compiler did not approve of using the 4-arg constructor of class CsvOutputFormat.
```

In the 0.5-SNAPSHOT version, we introduced an additional way to define the fields in the CSV input.

```
WordCount.TokenizeLine's ancestor AbstractFunction implements Seralized
```

But only in the 0.5-SNAPSHOT version. @uce made a pull request recently to introduce that.

Second set of questions:
1. No, you should have all dependencies by adding all jars in lib/.
2. If you had imported the "stratosphere-examples" project as a maven project into Eclipse maven would have downloaded the 0.5-SNAPSHOT dependencies into your local Maven repository and created a correct CLASSPATH for you. 
3. No. The easiest way is to open a shell and enter

``` bash
 curl https://raw.github.com/stratosphere/stratosphere-quickstart/master/quickstart.sh | bash
```

 and then import the contents of the quickstart/ directory into eclipse (or intellij) as a maven project.

**Comment by nullusDefectus** [comment]
It might very well be I have version differences between the binaries and the source. FWIW, I followed the instructions at http://stratosphere.eu/downloads/ to get both the Ready to Run package and the source. I should have known better that I was getting the current snapshot.

Regardless, I am still confused by your apparent insistence on using/needing Maven to view/build/run the example source. Modulo the version problem, it looks like my trial establishes that Maven is not necessary. That is, including source code with the Ready to Run Package, and may be including some simple instructions, makes it extremely easy for users to taste test the examples

Am I right or am I missing something?

**Comment by rmetzger** [comment]
How exactly did you get the source? We do not have instructions on getting
the source for the release on the Downloads page.

The thing with Maven is not easy to understand, but let me try:

You (need) maven to*:
- build Stratosphere from the source code
- to build the code of a Stratosphere program you've written
- view the source code without stuff being marked as errors

*You can do everything manually (without Maven) if you want, but that would
be crazy. As I said before, you usually have Maven.

You do not need maven to:
- run Stratosphere from the ready to run packages
- to run the examples in the ready to run packages

On Wed, Mar 5, 2014 at 9:54 PM, nullusDefectus notifications@github.comwrote:

> It might very well be I have version differences between the binaries and
> the source. FWIW, I followed the instructions at
> http://stratosphere.eu/downloads/ to get both the Ready to Run package
> and the source. I should have known better that I was getting the current
> snapshot.
> 
> Regardless, I am still confused by your apparent insistence on
> using/needing Maven to view/build/run the example source. Modulo the
> version problem, it looks like my trial establishes that Maven is not
> necessary. That is, including source code with the Ready to Run Package,
> and may be including some simple instructions, makes it extremely easy for
> users to taste test the examples
> 
> Am I right or am I missing something?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/stratosphere/stratosphere/pull/520#issuecomment-36792155
> .

## 

Robert Metzger, Kontakt: metzgerr@web.de, Mobil: 0171/7424461

**Comment by nullusDefectus** [comment]
Looks like we were both commenting simultaneously.

It looks like you and I are in agreement for the most part, except I don't think Maven needs to be even mentioned in the context of the Ready to Run package: may be mention it as an aside; definitely not so prominently.

**Comment by nullusDefectus** [comment]
http://stratosphere.eu/downloads/#source shows how to get the source code (without qualifying the version). The instructions there are simple and they work. It is just that I should have realized I might be getting code for a later build.

<pre>The thing with Maven is not easy to understand</pre> My point exactly.
<pre>You can do everything manually (without Maven) if you want, but that would be crazy. As I said before, you usually have Maven.</pre>

I disagree with both points. It is a lot easier to create an Eclipse project and add reference to lib than downloading and learning yet another tool. I wouldn't assume people _usually_ have Maven.

Bottom line, I think it is great if Ready to Run Package includes example source and if users are able to view/build/run in their favorite IDE without installing anything else. It looks like that will be possible with the next stable release.

**Comment by fhueske** [comment]
Hi @nullusDefectus,
First of all, thanks a lot for spending so much time with us, trying out the project, and giving lots of valuable feedback!

I also think, that the "Ready to Run" package should contain everything to run, view, and import  the examples into an IDE. We could build a example-source.jar (or zip) file that contains the source code of the examples and can be imported into an IDE _and_ have the plain source code files in the examples folder for instant inspection with an editor.

To get the code in the IDE, one would follow the steps you did before:
1. create a new project
2. import code from examples-source.jar
3. set build path to ./lib and ./lib-clients

A side from the examples-source.jar, this should already work as @rmetzger said.
So, there is no Maven required for this. In fact Maven is not mentioned in the context of the "Ready to Run" section of the Downloads page.

What I get out of this conversation is, that the Downloads page is mainly targeted to the group of "System Admins", i.e., people who want to set-up the system in a cluster and less for users that want to inspect the system by digging into the examples and running them from IDE or shell. 
For these guys, we have the QuickStart section of our page (although the aspect of getting the code of the examples in your IDE is not really covered there either). 
I think that many people from the users or "Developers" group will have look at Downloads first where they might not find the information they are looking for.

In general, we should make sure that people find the information they are looking for on our website (even if they do not know how we intended the website to be used ;-) ).
So we should make sure that the interests of all users are met by the Downloads page, which is  the forth most visited page (after start, quickstart, and main-docs-page).

**Comment by StephanEwen** [comment]
We could have a prepared eclipse project that includes sources and required libs and is pre-configured. Then you can simply open that project in eclipse and start browsing. No additional tools required, no setup or cofiguration.

**Comment by StephanEwen** [comment]
The cleaner solution for any developer that wants to write serious applications is definitely Maven (or some other dependendy managemet tool), but if that is considered a blocker for getting started or browsing, the prepared eclipse project may be a solution?

**Comment by rmetzger** [comment]
It is possible to generate the Eclipse configuration file using Maven. However, I don't know if it is possible to create "shipable" maven projects.

**Comment by fhueske** [comment]
Not everybody is using Eclipse. 
Including an importable archive with the examples code in the binary distribution is the most accessible solution IMHO.

**Comment by nullusDefectus** [comment]
@fhueske: I think you have nailed the issue in your first comment here. BTW, the QuickStart guide requires Maven, and Maven makes quick and prominent appearance in this repository in pretty much any conversation on using Stratosphere. Both of these need to be addressed. Of course, improved documentation goes a long way.

@StephanEwen: I totally agree that any one developing complex projects benefits from using Maven, but trying out examples hardly qualifies. Your proposal of including a prepared Eclipse project might work (if that is feasible) assuming Eclipse is the more popular IDE.

Here is what I feel is necessary and sufficient for a beginner:
1. Include example source in the Ready to Run package. I believe this is already done.
2. No need for an examples jar. While the jar makes it easy to import all examples at once, it adds work if user wants to examine examples selectively/progressively.
3. Add a Quickstart secion such as "Browse examples in IDE" after the section "Run Example" (http://stratosphere.eu/quickstart/setup.html)
   -- Use Eclipse as an example, or may be no need to call out any particular IDE.
   -- Just three steps: Create an Eclipse project; Add Stratosphere libraries to project; Add WordCount.Java to project.
   -- See below for thoughts on organizing example source
4. Link to Quickstart at the end of the Ready to Run Package section in the Downloads page.

Thoughts on organizing example source (Achtung: I have looked at only a few of the examples as of now, so these thoughts might not be ripe yet.)

It is unclear what "record" represents in the package hierarchy for examples. To my untrained eyes, it seems irrelevant at least to the wordcount examples. Perhaps this level can be removed.

I assume the examples illustrate using Stratosphere without Hadoop (WordCount), with Hadoop 1, with Hadoop 2, and YARN. It will be helpful to organize the examples along these lines. So, instead of "record", use intermediates such as "nohadoop", "hadoop1", "hadoop2", and "yarn". This proposed organization offers the following benefits:
1. Reflects the purpose of the examples.
2. Lines up with the Ready to Run Package user has downloaded. 
3. Helps include examples relevant only to the version downloaded, if you choose to.
4. Helps user work with increasing/varying complexity of examples. For example, user will know which examples need Hadoop installed.

Lastly, place each independent example in its own package/directory. In the current structure, I see  at least one package with multiple "programs", that is .java files with main method. For example, the wordcount package includes three such .java files. The file AnonymousWordCount.java has no main method and this class is not referenced in any of the other classes in the package. 

The pagerank package has 12 .java files, none of which have a main method. [Wipes glasses, checks again.]

My rule of thumb for examples: Unless the directory contains shared classes re-used across examples, each directory includes one and only .java file with a main method. I realize this approach increases the number of directories, but the user experience will be consistent. Alternatively, if a directory must contain multiple .java files with main method, use the suffix "Program" for the name of class/file that contains a main method and mention this pattern in QuickStart.

**Comment by fhueske** [comment]
You are definitely right. Our example programs are in a messed up state right now. Over time, examples were added and modified and things became inconsistent.
Cleaning up the examples module is an issue, which I find very important. The structure / organization you suggested of having a single java file and putting all other classes (if necessary) in a sub-package sounds like a very good idea. I will open an issue with detailed information for this right now.

A few side notes:
- Stratosphere programs are totally independent on whether Hadoop is used or in which version Hadoop is used. So there is no need to differentiate there.
- The `record` package differentiates between the old (Record-based) and the new (more generic) Java API. The new API is currently under construction and will be promoted as the default API once it is ready.

I also support your suggestions for the website.

Thanks for the good input!

**Comment by nullusDefectus** [comment]
Clarifying my thumb rule for examples: It is ok if a package has multiple Java files, but one and only of the files should contain the main method. A trivial/simple example such as wordcount might have just one file in the package; a non-trivial example (pagerank?) might end up having multiple files, even sup-packages.

Please consider this clarification in the context of the discussion on Issue #538.

Here is my poor attempt to illustrate the package organization.

examples
-- common [files common to all examples]
-- example1
---- example1.java [has main method}
-- example2
---- example2.java [has main method]
---- e2support1.java [does not have main method; example2.java references something in this file]
-- example3
---- example3_1
-------- example3_1.java [has main method]
---- example3_2
-------- example3_2.java [has main method]
-------- e312.java [does not have main method; example3_2.java references something in this file]
---- common
-------- [files common to example3_1 and example3_2; none of these have main method]

**Comment by StephanEwen** [comment]
@rmetzger Is this pull request ready?

**Comment by rmetzger** [comment]
The pull request is as described in my initial issue. (So if you want that, you can merge it)
It does not reflect the results of the discussion (the source files are not "import"-able into eclipse).

**Comment by StephanEwen** [comment]
Merged current version in 3b3260a6d7013fc8afb4fa9ac9744205aed50f81
